roxie/roxie.md

```diff
@@ -0,0 +1,155 @@
+be indicative of a query that could use some tuning)
+
+Before requesting rows from an input, it should be started, and when no more rows are
+required it should be stopped. It should be reset before destruction or reuse (for example
+for the next row in a child query).
+
+Balancing the desire to reduce latency with the desire to avoid wasted work can be tricky.
+Conditional activities (IF etc) will not start their unused inputs, so that queries can be
+written that do different index reads depending on the input. There is also the concept of a
+“delayed start” activity – I would need to look at the code to remind myself of how those are
+used.
+
+## Where are the Dragons?
+
+Splitter activities are a bit painful – they may result in arbitrary buffering of the data
+consumed by one output until another output is ready to request a row. It’s particularly
+complex when some of the outputs don’t start at all – the splitter needs to keep track of how
+many of the inputs have been started and stopped (an input that is not going to be used must
+be stopped, so that splitters know not to keep data for them). Tracking these
+start/stop/reset calls accurately is very important otherwise you can end up with weird bugs
+including potential crashes when activities are destroyed. Therefore we report errors if the
+counts don’t tally properly at the end of a query – but working out where a call was missed
+is often not trivial. Usually it’s because of an exception thrown from an unexpected place,
+e.g. midway through starting.
+
+Note that row requests from the activities above a splitter may execute on whichever thread
+downstream from the splitter happens to need that particular row first.
+
+The splitter code for tracking whether any of the downstream activities still need a row is a
+bit hairy/inefficient, IIRC. There may be scope to optimize (but I would recommend adding
+some good unit test cases first!)
+
+## How does “I beat you to it” work?
+
+When there are multiple agents fulfilling data on a channel, work is shared among them via a
+hash of the packet header, which is used to determine which agent should work on that packet.
+However, if it doesn’t start working on it within a short period (either because the node is
+down, or because it is too busy on other in-flight requests), then another node may take over. The IBYTI messages are used to indicate that a node has started to work on a packet and
+therefore there is no need for a secondary to take over.
+
+The priority of agents as determined by the packet hash is also used to determine how to
+proceed if an IBYTI is received after starting to work on a request. If the IBYTI is from a lower priority buddy (sub-channel) then it is ignored, if it’s from a higher priority one
+then the processing will be abandoned.
+
+When multicast is enabled, the IBYTI is sent on the same multicast channel as the original
+packet (and care is needed to ignore ones sent by yourself). Otherwise it is sent to all
```

Comment:
Otherwise is it sent to all buddy IPs ...
Not totally clear to me, should be an 'other' added ?
ie if multicast then it is sent to all and we have to be careful to ignore own sends, but if not multicast then msgs sent to all _other_ buddy IPs.

Reply:
If I read it again I think buddies implies all others except myself.